Pokedex branch 2 async

Sebs5384 opened the pull request:
Replace .then() with async/await for better readability in our code.

Sebs5384 committed:
Initializes main files

We initialize by creating all the main files

Sebs5384 committed:
Add main structure

This will be the main structure of our project a card that'll be later populated by more cards, a navbar and the footer with the paginator

Sebs5384 committed:
Add image

These image are going to be used background and nav banner

Sebs5384 committed:
Update styles

New styles for the buttons and paginator

Sebs5384 committed:
Add pokebutton animation

This will cause the effect of the pokeball shaking upon click to resemble a pokeming being catch

Sebs5384 committed:
Update html

This will now display a button resembling a pokeball on the navbar and the place holders for the cards that will be displayed as pokemons from the API

Sebs5384 committed:
Creates source files

These files will be the one holding all the logic of the page, if future implementations are needed, more files will be added.

Sebs5384 committed:
Add api functions

These functions will serve us to get a single pokemon information and a list of pokemons, in the future we'll work towards retrieving what we need of these with a function that takes such responsability

Sebs5384 committed:
Update apiPokemon

This introduces a new function that will be in charge of returning the pagination data we'll use later to create our pagination navbar

Sebs5384 committed:
Update html

This removes the paginator holder, we'll be settling it through a function that creates as many items as pokemons divided by the amount we want to show through screen

Sebs5384 committed:
Creates pagination

This function is in charge of creating the pagination, we also adding a previos and next button by the end of each side of the paginator

Sebs5384 committed:
Update styles

We'll use this during the dev. of our app

Sebs5384 committed:
Creates pagination functions

This will handle the paginator UI, if creates then hides all elements except for the previous and next button and will always display atleast 5 buttons upon navigating through it

Sebs5384 committed:
Update pagination

setup the pagination

Sebs5384 committed:
Update index

The sets the function that will setup the pokedex

Sebs5384 committed:
Creates setupPokedex

This function will be in charge of all the functions the pokedex need to start running

Sebs5384 committed:
Update pagination

This will give a more descritpive name to variables and parameters

Sebs5384 committed:
Update pagination

This will handle the logic of PREVIOUS and NEXT button in our paginator

Sebs5384 committed:
Fixes previous button

This fixes a bug where the previous button wouldn't go backwards since we were passing -1 rather than -2

Sebs5384 committed:
Update pagination

This will disable the previous and next button accordingly, either by reaching the last page or the first page.

It also updates createPaginator function, now it returns instead of cleaning the paginator.

Sebs5384 committed:
Update getPokemons

This will remove the next and previous since we are not in need of this.

Sebs5384 committed:
Update getPokemons

This will return TOTAL_POKEMONS instead of count

Sebs5384 committed:
Creates calculatePaginationValues

This function will be in charge of returning the paginator values

Sebs5384 committed:
Implements paginatorValues

The pagination values will be now returned by the function with alias pokedexPaginationValues

Sebs5384 committed:
Update api/pokemon.js

This will remove getPokemonPagination data, the reason we are doing this is because we'll use mappers later to organize it better, we rather using getPokemons for this and then map through the data it returns later to use the data as parameters of our functions

Sebs5384 committed:
Change variables

These variables were named on upperCase and snake case before, this is only being done when the variable is hard-coded now.

Sebs5384 committed:
Update pokedex

We are now handling all the modules in pokemon.js UI

Sebs5384 committed:
Update pagination

gives the pagination function a more descriptive name as well as the callback parameter

Sebs5384 committed:
Changes pokemon

This will move the module outside the folder UI as it will nothing to do with the user interface

Sebs5384 committed:
Changes calculatePaginationValues

This will be now exported and handle as an alias in the module that requires it

Sebs5384 committed:
Add getPokemonNames/Ids

These functions will be in charge of parsing the objects coming from our API call

Sebs5384 committed:
Removes html placeholders

These were using to have a view of how our pokedex would look like

Sebs5384 committed:
Add getPokemonSprite

This is used to return an png from the internet by giving it an ID

Sebs5384 committed:
Update cards

We making the structure of the cards that will be displayed upon the program running

Sebs5384 committed:
Add displayPokemonCards

This will be the function that display the pokemon name, id and img in a card

Sebs5384 committed:
Add toUpperCase function

This will parse our pokemon names to upperCase on its first letter of each phrase

Sebs5384 committed:
Handles 404 not found on image

The card sprites that returns 404 as response will now display a new image that is fixed in our img folder

Sebs5384 committed:
Add image

We'll use a removed bg for the 404 and a pokeball img for the cards bg

Sebs5384 committed:
Update cards

This will add the bg image to the cards

Sebs5384 committed:
Change firstLetterToUpperCase

Now its parsePokemonName and it removes the - on names too

Sebs5384 committed:
Creates general ui

This module is for general purposes.

Sebs5384 committed:
Add scale effect

This effect will have place when the user hovers around the pokemoin imgs

Sebs5384 committed:
Update pokedex

Setting the modal structure and logic on our pokemon.js file

Sebs5384 committed:
Add dummy modal

This dummy modal is the structure where we'll handle the cards that will shown upon clicking the pokemons

Sebs5384 committed:
Update HTML

Working on the dummy modal of our cards

Sebs5384 committed:
Add dummy imgs

We are currently using these to build or modal

Sebs5384 committed:
Update HTML

Currently working on the modal that'll be displayed as our pokemon card

Sebs5384 committed:
Update modal

This add the pokemon description, weakness and other characteristics of the  modal

Sebs5384 committed:
Update modal borders

Currently working on the styles our modal will have before adding the logic

Sebs5384 committed:
Add description icons

These will be placed on the pokemon's status and skills,

Sebs5384 committed:
Update html

We are done with the structure of our template for our modal, we'll proceed to remove the styles and store them in css

Sebs5384 committed:
Add icons

These icons are being added to structure our template, but later will be added to water pokemon types

Sebs5384 committed:
Update img folder

This will create a folder to organize our imgs and icon

Sebs5384 committed:
Update styles

We are removing the styles we were using to structure our modal and now making them css classes

Sebs5384 committed:
Update API module

This will add a new api call which is needed to display information about the pokemon in our modal, we also adding .catch to our api calls

Sebs5384 committed:
Creates handleClickedPokemon

This function will return the ID of the clicked pokemon, it will later be a callback function for our modal

Sebs5384 committed:
Changes styles

These styles will be now have different names that makes more sense now on what they represent

Sebs5384 committed:
Updates HTML

We are giving the modal sections ids and moving some classes where they make more sense

Sebs5384 committed:
Creates setupPokemonModal

This is the main function in our modal module, it will orchestrate the creating of the modal, display and anything that has something to do with the modal.

Sebs5384 committed:
Update pokedex

We'll setting setupPokemonModal in here for the time being while we work towards its whole implementation.

Sebs5384 committed:
Update pokedex

This will create two functions that return a promise with an object with the values we need in the main pokedex file, this is being done to process all the data before using it

Sebs5384 committed:
Update img

This will add all the types logo, we are using these to display them in the pokemon modal accordnig to the pokemon type

Sebs5384 committed:
Update styles

This will add a few changes to the styles that already exists

Sebs5384 committed:
Updates modal

Modal now will generate its content through JS, this is being done to avoid loading it without the user using it, it will be created and later removed when its not needed anymore in our HTML, changes to it will be done later.

Sebs5384 committed:
Add all type icons

We'll use this in our modal to have personalized icons depending on the type of pokemon the user choose

Sebs5384 committed:
Updates modal

This will set the right path to the icons/logos

Sebs5384 committed:
Add textures

These are the textures for our modal, we'll be using them depending on the pokemon that is being displayed taking as reference its type

Sebs5384 committed:
Update HTML

This will remove the dummy modal we were using to structure it

Sebs5384 committed:
Add type styles

This will add the background of each type of pokemon, respectively it will be assigned when the user interacts with the modal

Sebs5384 committed:
Updates modal

This will give the parameter needed to display a background while we work on it

Sebs5384 committed:
Update modal

This will separate the creation / display and texture change from the function that should only create the modal.

Sebs5384 committed:
Update general

This will create two functions that will take care of parsing the height and weight to lbs and feet

Sebs5384 committed:
Cleans up modal

We are currently working towards the cleaning of the modal attributes and how they are being pass

Sebs5384 committed:
Update css

Moving all the .type linear gradients to the bottom of css since they are not part of the main styles

Sebs5384 committed:
Update modal

Cleaning up the way the attributes are being pass down from the function that proccesses them

Sebs5384 committed:
Cleans up modal

Currently working on the clean up of data processing

Sebs5384 committed:
Add advantage

This will display weakness and resistance types for the selected pokemon

Sebs5384 committed:
Update utils

Utils now has a new module pokemon.js which will be the one making the process of data by using the functions in general.js

Sebs5384 committed:
Update misc stats

This will add a condition checking if this is a basic pokemon then to display 1 retreat icon only

Sebs5384 committed:
Add small text-shadow

We are adding a small text-shadow to the skills font to prevent it not being visible when its a fire type pokemon

Sebs5384 committed:
Replace uni-code

This will replace the uni-code that comes from the pokemon description with a space, otherwise the text will look off

Sebs5384 committed:
Fixes genus bug

This will fix an small bug that caused the new generations to never open in the modal, due to the specific [7] index in the object, now we are passing down a find method to bring down the first english language it finds

Sebs5384 committed:
Update modal

This will check for the genus being empty

Sebs5384 committed:
Update cards

This will display the new GEN cards with the correct number of ID

Sebs5384 committed:
Resolve promises separately

We now solve promises in this way to make use of the genus information, we now call species with this name to avoid a bug that caused to make imposible to click new GEN pokemons

Sebs5384 committed:
Creates getPokemonMainName

This function will display the pokemon first name inside the modal

Sebs5384 committed:
Updates HTML

This will create the structure for our search input that is used to jump to the desired page

Sebs5384 committed:
Creates validation module

We'll create all the code related to validations in this file

Sebs5384 committed:
Creates page search logic

This creates the logic for our search input

Sebs5384 committed:
Update validations

This will display a popup if conditions inside the validations are met, we'll display the error in screen.

Sebs5384 committed:
Add validations

Fixing a small bug that allowed the user to go to page 0 and 66

Sebs5384 committed:
Update styles

This will add a border to the popover and also make it round

Sebs5384 committed:
Add pokemon font

This will add the ttf for emerald pokemon version that we'll use in the catch-pokemon feature

Sebs5384 committed:
Update styles

organizing and cleaning up a bit

Sebs5384 committed:
Update HTML

This will add the modal structure container that we are going to display if the user has caught a pokemon, it also adds its animation and styles

Sebs5384 committed:
Update general UI

This function is being made in order to handle the button that will trigger the caught pokemon modal, we also created a function that changes the text of the modal according to the caught pokemon name

Sebs5384 committed:
Update modal

We are creating the modal that is going to be trigger by the catch pokemon button, we also updated show modal and created hideModal

Sebs5384 committed:
Creates getCaughtPokemonData

This function is responsible for processing the data that will be used in our modal, we also created a function to randomize the pokemon caught, this will allow us to get a different pokemon to be displayed in our modal everytime we click the catch button

Sebs5384 committed:
Set setupCatchPokemon

Settling the function that will handle the catchPokemon events

Sebs5384 committed:
Update HTML

This updates the navbar, we now divide it for a better structure

Sebs5384 committed:
Creates registrationModal

This will be displayed after the caught pokemon modal, it resembles a emerald-pokedex registration, we'll store the registrated pokemon next to the pokeball

Sebs5384 committed:
Add registrationModal styles

This will add the styles needed to make our modal alike the real one

Sebs5384 committed:
Add registrationModal

This will add the registration function to the entry point

Sebs5384 committed:
Add pokeball delay

This will disable the button upon clicking it in order to avoid the user from spamming it

Sebs5384 committed:
Update getPokemonSprite

It now accepts a new parameter, this being the desired artwork / sprite we wish to get

Sebs5384 committed:
Update img folder

Deleting unused imgs and add pokedex len img

Sebs5384 committed:
Update HTML

Removing the default styles of imgs and adding them into a class, we are also adding static backdrop to the caught modal

Sebs5384 committed:
Update modal

This changes how the data that the modals uses are given, now each modal section will require the specific parameters that they will use

Sebs5384 committed:
Creates setCaughtPokemonSlot

This function will be in charge of setting down the caught pokemons through our "catch a pokemon" mini game, it will store as an IMG in our HTML, we are also displaying the card modal upon clicking on the desired slot

Sebs5384 committed:
Update pokedex

Sebs5384 committed:
Create catch-pokemon module

This will hold all the logic related to the catch-pokemon mini-game

Sebs5384 committed:
Creates search-box module

This holds the logic of our search-box feature

Sebs5384 committed:
Settles search-box feature

This will implement the search-box feature

Sebs5384 committed:
Update pokemon

Sebs5384 committed:
Add searchBox style

This will display an scroll effect on the searchBox

Sebs5384 committed:
Update handleSearchInput

This will handle now the enter event, whenever the user press enter on the pokemon name we'll check if its exist and if it exist and matches the name input by the user then we'll use the ID

Sebs5384 committed:
Update searc box

This implements the last feature, showing the modal according to the input of the user

Sebs5384 committed:
Update pagination

This removes the responsability of getPageData of retrieving the pokemonData as well, we now divided it in two functions, this will serve us well to refactor the navigation bar

Sebs5384 committed:
Update pokedex

This update handles the pokemonSprites in another way

Sebs5384 committed:
Delete getPokemonsData and getCaughtPokemonData

Since we divided getPokemonsData responsabilities with getPageData now we can use it instead of getPokemonsData and getCaughtPokemonsData which were doing the same but with extra-steps

Sebs5384 committed:
Delete handleSearchBoxClick

This will not be used anymore, since handleClickedPokemon now relies on the whole DOM rather than just focusing one element, this means that we'll be able to click whatever pokemon as long as it has data-id

Sebs5384 committed:
Update pokedex structure

This will now make an API call for the whole nav-bar, this is being done for two reasons, one optimize the api calls and two, not being dependent of multiple GET-SOMETHING function, i decided that this will be better than making 3 api calls to limit 10000 to fill the three pokemon slots and then another to create the list, instead now we use the list length to randomize the caught pokemon and then make a single pokemon api call to bring down the data.

Sebs5384 committed:
Update cards

This will add a more descriptive logic of what we are trying to imply in here

Sebs5384 committed:
Update modal

This introduces removeModals function, we were having an issue where if we opened multiple modals and then pressed the card close button it would only close one instance, now we remove them all at once when pressing it by using a function

Sebs5384 committed:
Add responsive

This makes the main HTML responsive

Sebs5384 committed:
Add CDN

We adding the CDN to deploy the project with bootstrap bundles

Sebs5384 committed:
Update BG url paths

Sebs5384 committed:
Updates css

Sebs5384 committed:
Path texture

This will path the texture to be display correctly in our deployment

Sebs5384 committed:
Updates footer

This removes the extra height in our app, which was being cause by the component "container" from bootstrap

Sebs5384 committed:
Add responsive to modal

Sebs5384 committed:
Update responsive design

Adding more responsive design for 420px

Sebs5384 committed:
Add registration modal responsive

Adjust the responsive design of the two other modals in 420px

Sebs5384 committed:
Updates responsive design

This will take care of 920px screen-width

Sebs5384 committed:
Update responsive

This updates screen-width 800px

Sebs5384 committed:
Add responsive design

This finished the caughtPokemon modals in 800px

Sebs5384 committed:
Changes shadow

Now setting it in the css class

Sebs5384 committed:
Update cards

This will add a better style to our cards

Sebs5384 committed:
Fix 800px pagination searcher

This displays the paginator search box correclty

Sebs5384 committed:
Add pokedex fixtures

These are goign to be use in our test case in cypress to stub the responses from the API

Sebs5384 committed:
Add data-cy

We'll use these attributes to make mention of the elements in our test case

Sebs5384 committed:
Add pagination fixtures

This will be useful for our pagination stub in cypress test case

Sebs5384 committed:
Add pokedex cy test

This will add the test cases for the pokedex

Sebs5384 committed:
Update README.md

Updates read me

Sebs5384 committed:
Update README.md

Update reads me

Sebs5384 committed:
Revert "Update pokedex"

This reverts commit 6d85aea8c6ad084ec6258380f9a23ba4528bcdfa.

Sebs5384 committed:
Update pokedex

Working on async functions

Sebs5384 committed:
Update styles

We are now adding a fixed position for the background and fixed height for the cards too, this is being done in order to remove the background repeating itself and then the cards-container background to minimize and maximize when using the paginator

Sebs5384 committed:
Update modals

Adding async to the functions in charge of loading the modal conent

Sebs5384 committed:
Add small delay to modal

This will add a 0.5 second delay to the pokemon modal before showing it

Sebs5384 committed:
Update changeModalTexture

Testing out the onload event in order to load the texture before showing the modal

Sebs5384 committed:
Changes promises to async

We are now resolving our promises by defining our functions as async, this way we can replace .then when resolving them

Sebs5384 committed:
Removes background-img

We are now using linear-gradient instead.

Sebs5384 committed:
Update modal sprite

This changes how the modal is actually displaying the sprite in our modal, now it awaits until it has been loaded before showing the modal.

Sebs5384 committed:
Update modal

Now showModal and hideModal awaits for a new Promise object to solve before doing their respective functionalities

Sebs5384 committed:
Creates displayLoadingModal

This will set a loading message when the user awaits for the pokemonModal to load all the textures and sprites

Sebs5384 committed:
Update modal

Sebs5384 committed:
Delays hideModal

This will give enough time for hideModal to process its execution before showing the new modal

Sebs5384 committed:
Changes loadingModal

This will clean the modal upon creating a new one

Sebs5384 committed:
Create loadCardsSprite/background

This will await for the sprites and background to load before showing the cards at the begin of the program and when navigatin through the paginator

Sebs5384 committed:
Add error event to modal

On error will now be displayed ONLY when the card doesn't have an sprite, rather than when changing it with the previous stage icon

Sebs5384 committed:
Update caughtPokemon sprite

We'll bring the default sprite for the registration modal instead to attach to its model in the pokemon game

Sebs5384 committed:
Update loadingModal

it now hides the pokemon card modal while its loading, this is being done to fix an small bad UI experience where the user could see the card loading in the background when click previous evolution

Sebs5384 committed:
Fix small UI bug on mobile

This fixes bad experience in UI mobile, which consisted in the loading message displaying way too big since its min-height was being shared with the registration text as well

GonzaloAlt commented on cypress/e2e/pokedex.cy.js:
```diff
@@ -0,0 +1,124 @@
+    cy.get("[data-cy='page-4']").should("be.visible");
+  });
+
+  it("Uses the pokemon search box", () => {
+    cy.intercept("GET", blastoiseURL, { fixture: "blastoise" }).as("getBlastoise");
+    cy.intercept("GET", blastoiseSpeciesURL, { fixture: "blastoiseSpecies" }).as("getBlastoiseSpecies");
+
+    cy.get("[data-cy='pokemon-search-box']").click();
+    cy.get("[data-cy='pokemon-search-box-list']").children().should("have.length", 1292);
+    cy.get("[data-cy='pokemon-search-box-list']").should("be.visible");
+
+    cy.get("[data-cy='pokemon-search-box']").type("Blastoise");
+    cy.get("[data-cy='Blastoise']").should("be.visible");
+
+    cy.get("[data-cy='Blastoise']").click();
+    cy.get("[data-cy='pokemon-card-modal']").should("be.visible");
+
+    cy.wait(1000);
+    cy.get("[data-cy='pokemon-modal-close-button']").click();
+
+    cy.get("[data-cy='pokemon-search-box-list']").should("not.be.visible");
+
+    cy.get("[data-cy='pokemon-search-box']").click();
+    cy.get("[data-cy='pokemon-search-box-list']").should("be.visible");
+    cy.get("[data-cy='pokemon-search-box']").clear();
+    cy.get("[data-cy='pokemon-search-box']").type("Wartortle");
+
+    cy.get("[data-cy='Wartortle']").should("be.visible");
+  });
```
Maybe you should isolate the tests a little bit more? One to search, another for closing the modal, etc.
Also, what is `1292`? can you save it in a variable? And could the title of the tests be more descriptive? `it('should do/show/etc something...', ()=>{})`

Sebs5384 replied:
I gave it a better description, isolated the test cases where the modal should respond to the close button and when the user uses the paginator search-box feature.

GonzaloAlt commented on cypress/e2e/pokedex.cy.js:
```diff
@@ -0,0 +1,124 @@
+    cy.get("[data-cy='Wartortle']").should("be.visible");
+  });
+
+  it("Clicks on pokemon card", () => {
```
Can you put something more descriptive?

Sebs5384 replied:
Added your feedback to my spec file, thank you !

GonzaloAlt commented on cypress/e2e/pokedex.cy.js:
```diff
@@ -0,0 +1,124 @@
+    cy.get("[data-cy='pokemon-modal-close-button']").click();
+  });
+
+  it("Uses paginator", () => {
```
Same here.

Sebs5384 replied:
Added it in here as well, thanks for the feedback !!

Sebs5384 committed:
Update cypress spec

This will add a better description and isolation to our test cases

Sebs5384 committed:
Add eslint

We are adding eslint to our project to make our code more clean

Sebs5384 committed:
Refactor pokedex

This will refact the whole pokedex logic by following Eslint practices

Sebs5384 committed:
Refactor cypress spec

Following eslint practices in the cypress spec file, it now follow said practices

Sebs5384 committed:
Update quotes

Using single quotes now instead of double

Sebs5384 committed:
Create pokemon storage

This file contains the local storage logic

Sebs5384 committed:
Create service

This creates the logic of the service we will offer to our application.

Sebs5384 committed:
Changes getSprite

We changing back the way we were requesting sprites with our function

Sebs5384 committed:
Add sprites to service

We will avoid using the API module in our pokedex file this way, instead we are going to offer it our service

Sebs5384 committed:
Add npm-run-all

Adding this into dev-dependencies, it will help us as a short-cut when running programs

Sebs5384 committed:
Installs jest and babel

We'll be using jest to run our unit tests

Sebs5384 committed:
Add getSpriteList

This function is bieng made in order to map through the pokemons that will be shown in each page

Sebs5384 committed:
Fixes sprite artwork

Now returns the official-artwork when requested instead of the classic one

Sebs5384 committed:
Update eslint

This will disable object destructuring on objects

Sebs5384 committed:
Fixes service pagination

Previously we were giving it a parameter that was never being use and that caused pagination to work oddly, it is now fixed
